File: src/org/stormgears/powerup/subsystems/information/RobotConfiguration.java
package org.stormgears.powerup.subsystems.information;

import org.stormgears.utils.configuration_file.AbstractRobotConfiguration;

public class RobotConfiguration extends AbstractRobotConfiguration {
	private static RobotConfiguration instance = new RobotConfiguration();

	public static RobotConfiguration getInstance() {
		return instance;
	}

	/*
	 * Public property fields: set these in loadExtras() so they can be updated from the config file.
	 * Remember, these will be set *once*. These values *will not* change throughout execution of the robot code.
	 */
	public boolean hasNavX;
	public int frontLeftTalonId, frontRightTalonId, rearLeftTalonId, rearRightTalonId;
	public double velocityF, velocityP, velocityI, velocityD;
	public double positionP, positionI, positionD;
	public int velocityIzone;
	public int positionIzone;
	public int encoderResolution;
	public boolean reverseJoystick;


	/*
	 * WARNING: Do not set properties here. Only access ones from the file and set their respective fields.
	 * If you want to set robot properties, SSH into the roboRIO and edit /home/lvuser/config.properties in vim.
	 * The whole point of having the file is to make properties _different_ on each physical robot.
	 *
	 * Also, parseInt() will return Integer.MIN_VALUE if it cannot successfully parse an int from the String provided
	 */
	@Override
	protected void loadExtras() {
		hasNavX = Boolean.parseBoolean(properties.getProperty("hasNavX"));
		frontLeftTalonId = parseInt(properties.getProperty("frontLeftTalonId"));
		frontRightTalonId = parseInt(properties.getProperty("frontRightTalonId"));
		rearLeftTalonId = parseInt(properties.getProperty("rearLeftTalonId"));
		rearRightTalonId = parseInt(properties.getProperty("rearRightTalonId"));

		velocityF = parseDouble(properties.getProperty("velocityF"));
		velocityP = parseDouble(properties.getProperty("velocityP"));
		velocityI = parseDouble(properties.getProperty("velocityI"));
		velocityD = parseDouble(properties.getProperty("velocityD"));
		velocityIzone = parseInt(properties.getProperty("velocityIzone"));

		positionP = parseDouble(properties.getProperty("positionP"));
		positionI = parseDouble(properties.getProperty("positionI"));
		positionD = parseDouble(properties.getProperty("positionD"));
		positionIzone = parseInt(properties.getProperty("positionIzone"));

		encoderResolution = parseInt(properties.getProperty("encoderResolution"));

		reverseJoystick = Boolean.parseBoolean(properties.getProperty("reverseJoystick"));
	}
}

File: src/main/java/org/stormgears/powerup/subsystems/information/RobotConfiguration.kt
package org.stormgears.powerup.subsystems.information

import org.stormgears.utils.configurationfile.BaseRobotConfiguration

object RobotConfiguration : BaseRobotConfiguration(useBackupIfFileNotAvailable = false) {

	/*
     * Public property fields: set these in loadExtras() so they can be updated from the config file.
     * Remember, these will be set *once*. These values *will not* change throughout execution of the robot code.
     */
	val robotWidth = getDouble("robotWidth")
	val robotLength = getDouble("robotLength")
	val robotHeight = getDouble("robotHeight")

	val hasNavX = getBoolean("hasNavX")
	val navXInvert = if (getBoolean("isNavXInverted")) -1.0 else 1.0
	val navXPort = getString("navXPort")

	val frontLeftTalonId = getInt("frontLeftTalonId")
	val frontRightTalonId = getInt("frontRightTalonId")
	val rearLeftTalonId = getInt("rearLeftTalonId")
	val rearRightTalonId = getInt("rearRightTalonId")

	val intakeLeftTalonId = getInt("intakeLeftTalonId")
	val intakeRightTalonId = getInt("intakeRightTalonId")
	val intakeArticulatorTalonId = getInt("intakeArticulatorTalonId")

	val elevatorMasterTalonId = getInt("elevatorMasterTalonId")
	val elevatorSlaveTalonId = getInt("elevatorSlaveTalonId")
	val sideshiftTalonId = getInt("sideshiftTalonId")
	val gripperTalonId = getInt("gripperTalonId")

	val velocityF = getDouble("velocityF")
	val velocityP = getDouble("velocityP")
	val velocityI = getDouble("velocityI")
	val velocityD = getDouble("velocityD")
	val velocityIzone = getInt("velocityIzone")

	val positionP = getDouble("positionP")
	val positionI = getDouble("positionI")
	val positionD = getDouble("positionD")
	val positionIzone = getInt("positionIzone")

	val elevatorRaiseP = getDouble("elevatorRaiseP")
	val elevatorRaiseI = getDouble("elevatorRaiseI")
	val elevatorRaiseD = getDouble("elevatorRaiseD")
	val elevatorLowerP = getDouble("elevatorLowerP")
	val elevatorLowerI = getDouble("elevatorLowerI")
	val elevatorLowerD = getDouble("elevatorLowerD")

	val elevatorStiffness = getDouble("elevatorStiffness", 1.0)

	val encoderResolution = getInt("encoderResolution")

	val reverseJoystick = getBoolean("reverseJoystick")

	val wheelRadius = getDouble("wheelRadius")

	val enableSensors = getBoolean("enableSensors", true)
	val enableDrive = getBoolean("enableDrive", true)
	val enableIntake = getBoolean("enableIntake", true)
	val enableElevator = getBoolean("enableElevator", true)
	val enableClimber = getBoolean("enableClimber", true)
	val enableGripper = getBoolean("enableGripper", true)

	init {
		if (navXPort != "USB" && navXPort != "MXP") {
			throw IllegalStateException("navXPort must be USB or MXP")
		}
	}
}
